Tests failing on macOS

Sometime between https://github.com/jaraco/keyring/actions/runs/1226964522 and https://github.com/jaraco/keyring/actions/runs/1373384293, CI tests started failing on macOS. ~~The tests pass for me locally on macOS Monterey, so something likely changed in the CI environment.~~

Error message is:

```
__________________________________________________________ ERROR collecting keyring/backends/macOS/api.py ___________________________________________________________
keyring/backends/macOS/api.py:45: in <module>
    SecItemAdd = _sec.SecItemAdd
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ctypes/__init__.py:377: in __getattr__
    func = self.__getitem__(name)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ctypes/__init__.py:382: in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
E   AttributeError: dlsym(RTLD_DEFAULT, SecItemAdd): symbol not found

```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,3 +1,14 @@
+v23.3.0
+-------
+
+* #529: macOS backend is no longer viable if the API module
+  cannot be loaded. Prevents "symbol not found" errors on
+  macOS 11 (Big Sur) and later when a "universal2" binary
+  is not used (available for Python 3.8.7 and later).
+
+* #547: Tests no longer attempt to run macOS backends even
+  on macOS when the backend is non-viable.
+
 v23.2.2
 -------
 

--- a/README.rst
+++ b/README.rst
@@ -54,6 +54,13 @@ install dbus-python as a system package.
 
 .. _dbus-python: https://gitlab.freedesktop.org/dbus/dbus-python
 
+Compatibility - macOS
+=====================
+
+macOS keychain support macOS 11 (Big Sur) and later requires Python 3.8.7
+or later with the "universal2" binary. See
+`#525 <https://github.com/jaraco/keyring/issues/525>`_ for details.
+
 Using Keyring
 =============
 

--- a/conftest.py
+++ b/conftest.py
@@ -1,5 +1,4 @@
-import sys
-import platform
+import ctypes
 
 collect_ignore = ["hook-keyring.backend.py"]
 
@@ -12,16 +11,11 @@ def macos_api_ignore():
     Ref #525.
     """
 
-    def make_ver(string):
-        return tuple(map(int, string.split('.')))
-
-    release, _, _ = platform.mac_ver()
-
-    return (
-        platform.system() != 'Darwin'
-        or make_ver(release) > (11,)
-        and sys.version_info < (3, 8, 7)
-    )
+    try:
+        ctypes.CDLL(ctypes.util.find_library('Security')).SecItemAdd
+        return False
+    except Exception:
+        return True
 
 
 collect_ignore.extend(['keyring/backends/macOS/api.py'] * macos_api_ignore())

--- a/conftest.py
+++ b/conftest.py
@@ -1,8 +1,29 @@
+import sys
 import platform
 
 collect_ignore = ["hook-keyring.backend.py"]
 
-if platform.system() != 'Darwin':
-    collect_ignore.append('keyring/backends/macOS/api.py')
+
+def macos_api_ignore():
+    """
+    Starting with macOS 11, the security API becomes
+    non-viable except on universal2 binaries.
+
+    Ref #525.
+    """
+
+    def make_ver(string):
+        return tuple(map(int, string.split('.')))
+
+    release, _, _ = platform.mac_ver()
+
+    return (
+        platform.system() != 'Darwin'
+        or make_ver(release) > (11,)
+        and sys.version_info < (3, 8, 7)
+    )
+
+
+collect_ignore.extend(['keyring/backends/macOS/api.py'] * macos_api_ignore())
 
 collect_ignore.append('keyring/devpi_client.py')

--- a/keyring/backends/macOS/__init__.py
+++ b/keyring/backends/macOS/__init__.py
@@ -28,6 +28,8 @@ class Keyring(KeyringBackend):
         """
         if platform.system() != 'Darwin':
             raise RuntimeError("macOS required")
+        if 'api' not in globals():
+            raise RuntimeError("Security API unavailable")
         return 5
 
     def set_password(self, service, username, password):

--- a/tests/backends/test_macOS.py
+++ b/tests/backends/test_macOS.py
@@ -1,16 +1,14 @@
-import sys
-
 import pytest
 
+import keyring
 from keyring.testing.backend import BackendBasicTests
 from keyring.backends import macOS
 
 
-def is_osx_keychain_supported():
-    return sys.platform in ('mac', 'darwin')
-
-
-@pytest.mark.skipif(not is_osx_keychain_supported(), reason="Needs macOS")
+@pytest.mark.skipif(
+    not keyring.backends.macOS.Keyring.viable,
+    reason="macOS backend not viable",
+)
 class TestOSXKeychain(BackendBasicTests):
     def init_keyring(self):
         return macOS.Keyring()

--- a/tests/backends/test_macOS.py
+++ b/tests/backends/test_macOS.py
@@ -9,6 +9,6 @@ from keyring.backends import macOS
     not keyring.backends.macOS.Keyring.viable,
     reason="macOS backend not viable",
 )
-class TestOSXKeychain(BackendBasicTests):
+class Test_macOSKeychain(BackendBasicTests):
     def init_keyring(self):
         return macOS.Keyring()

